### ISSUE TITLE
ssh fails to fetch if not all key types are supported

### ISSUE THREAD

**Issue by ehuss** [libgit2/libgit2#6454]
### Reproduction steps

Attempt to fetch an SSH repository from github.com with a version of ssh2 that is built on Windows with the WinCNG backend. You will need to have the 3 github.com keys in your known_hosts file (ssh-ed25519, ssh-rsa, ecdsa-sha2-nistp256), and importantly the ssh-rsa key should not be first.

### Expected behavior

Fetching should be successful.

Also, I feel like the order of the known_hosts keys should not be relevant. Perhaps this is a limitation of ssh2, but I believe this isn't an issue with OpenSSH.

### Actual behavior

The error is:

```
failed to set hostkey preference: The requested method(s) are not currently supported
```

The issue is that the WinCNG backend does not support ED25519 or ECDSA. The code that sets the key preference [here](https://github.com/libgit2/libgit2/blob/036fe1afba37d9b20509671d3048c6088f018134/src/libgit2/transports/ssh.c#L550-L558) does not know that.

Perhaps it could loop over the available keys until `libssh2_session_method_pref` doesn't fail?

### Workaround

A workaround is to use the OpenSSL backend.

### Version of libgit2 (release number or SHA1)

1.5.1

### Operating system(s) tested

Windows 10 Pro, Windows Server 2022

**Comment by ethomson** [comment]
Bluh. I feel like the deeper I look at this, the more terrible it becomes. If libssh2 doesn't support ED25519 or ECDSA (when built with WinCNG) what _are_ you using? RSA with a newer cipher? My understanding is that RSA only supports SHA1 in any released version -- are you using a version built out of main?

**Comment by carlosmn** [comment]
Yep, this looks like the way to solve it. Loop until we run out of known hosts or we find something that we can set. I hadn't thought of libssh2 not supporting the methods recorded in known_hosts. This is all pretty upsetting like Ed says. I haven't looked more into this, but I sure hope you can distinguish errors from not supporting a specific method.

**Comment by ehuss** [comment]
> If libssh2 doesn't support ED25519 or ECDSA (when built with WinCNG) what _are_ you using? RSA with a newer cipher? My understanding is that RSA only supports SHA1 in any released version

It depends on the version. A new release of libssh2 hasn't been made in a long time. We are moving to use an arbitrary version from master which has added rsa-sha2 support (https://github.com/libssh2/libssh2/commit/8b3a864c5bd69a72e7b14a2d94d57dd935aace2f). The 1.10 version of libssh2 WinCNG only supports `ssh-rsa` (the SHA1 version), and `ssh-dss` (which it probably shouldn't support).

I think one option is to iterate over all the `LIBSSH2_KNOWNHOST_KEY_*` types in `find_hostkey_preference()`, and collect all key types that return `LIBSSH2_KNOWNHOST_CHECK_MISMATCH`, and pass a comma-separated list to `libssh2_session_method_pref()`.

I might also suggest sorting those key types to match OpenSSH's preferred order, which I think is something like `ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-rsa`.

That would help ensure that the order of entries in known_hosts doesn't matter.

**Comment by ethomson** [comment]
> It depends on the version. A new release of libssh2 hasn't been made in a long time. We are moving to use an arbitrary version from master which has added rsa-sha2 support (https://github.com/libssh2/libssh2/commit/8b3a864c5bd69a72e7b14a2d94d57dd935aace2f). The 1.10 version of libssh2 WinCNG only supports ssh-rsa (the SHA1 version), and ssh-dss (which it probably shouldn't support).

Yeah, that's what I was wondering. Even with that, though, I'm struggling to make anything work. I replaced the libssh2 that we use in CI with that latest version in the repo, which does include 8b3a864 (I'm staring at the code now to make sure that it has that patch) and I'm still getting:

> `ERROR: You're using an RSA key with SHA-1, which is no longer allowed. Please use a newer client or a different key type.`

libssh2. 😡

**Comment by ehuss** [comment]
Is there a branch I could take a look at?

The important part is that the current `find_hostkey_preference` implementation needs to change. When it sees a host-key type of `ssh-rsa`, it needs to include in the preference list `"rsa-sha2-512,rsa-sha2-256,ssh-rsa"`.  Just specifying `ssh-rsa` as it currently does will prevent it from using the sha2 variant in the negotiation, and will likely result in that error.

**Comment by ethomson** [comment]
Yes - https://github.com/libgit2/libgit2/pull/6455

I cannot get libssh2 to talk to GitHub. I built the version in main and it seems unable to connect with an RSA key. I get the error about sunsetting SHA1.

This is with a known hosts file that only contains the RSA key. I haven't succeeded in connecting at all, either on my local machine or CI, so I haven't looked at the known hosts issue yet.

😓

**Comment by ehuss** [comment]
I think I see the issue with your CI setup. The tests are loading the `libssh2.dll` from the `/c/tools/php` directory (which is in PATH). In `test.sh` you'll need to add something like `PATH=$BUILD_TEMP/libssh2/bin` to pick up the one you are downloading.

I've been able to get the tests running locally and on GitHub Actions.

BTW, let me know if you want me to put together a suggested patch for creating the sorted-order key preferences. I think it should be pretty simple.

**Comment by ethomson** [comment]
> I think I see the issue with your CI setup. The tests are loading the libssh2.dll from the /c/tools/php directory (which is in PATH). In test.sh you'll need to add something like PATH=$BUILD_TEMP/libssh2/bin to pick up the one you are downloading.

Yes, thanks, that was the problem in CI. I refactored our build and test scripts -- we were mutating the path when running the build steps, but not the test, and we were doing it wrong on Windows. (We used `cygpath` on the `PATH` environment variable, which mangled things -- oops.) It looks like things are working properly now.

But I still can't get libssh2 to work with in-memory keys when built against WinCNG...  it's hard for me to debug on Windows but it looked like it was getting stuck in an infinite loop. Building libssh2 against OpenSSL _does_ seem to work. I'll see if I can put together a simple repro case for libssh2 later.

> BTW, let me know if you want me to put together a suggested patch for creating the sorted-order key preferences. I think it should be pretty simple.

That would be _most_ helpful and appreciated.

**Comment by ehuss** [comment]
This is what I'm thinking of for setting the hostkey preference: https://github.com/ehuss/libgit2/commit/cf6cbc4b750fe76001c05624a9b9f8efe9e3b514

Sorry I'm not too familiar with your test suite or coding standards, but hopefully that can be a helpful starting point.

That checks for each hostkey in order, and builds up a preference string. If the hostkey algorithm isn't supported, libssh2 will ignore it. If none of the hostkeys are configured, then no preference will be specified.

That should fix it so that the order of keys in known_hosts shouldn't matter.

**Comment by ehuss** [comment]
I'm going to close as effectively resolved by #6455.  libgit2 should now add all found key types to the host key preference list in a priority order, and libssh2 will ignore any that are not supported. Thanks @ethomson!

I'd still like to eventually see [`load_known_hosts`](https://github.com/libgit2/libgit2/blob/c058aa87dce4c67a3b86b3349beebd64b7bedcd3/src/libgit2/transports/ssh.c#L435) not treat loading failures as a hard error. For our use case, we have our own known-hosts management, and I'm a little concerned about the libssh2/libgit2 handling being a limiting factor. However, I don't believe libgit2 supports any sort of warning mechanism, and silently ignoring errors could be hard for others to diagnose. And I currently don't know of a concrete example that is currently a problem.
